plugin/discovery/find.go

```diff
@@ -0,0 +1,167 @@
+package discovery
+
+import (
+	"io/ioutil"
+	"path/filepath"
+	"runtime"
+	"strings"
+)
+
+const machineName = runtime.GOOS + "-" + runtime.GOARCH
+
+// FindPlugins looks in the given directories for files whose filenames
+// suggest that they are plugins of the given kind (e.g. "provider") and
+// returns a PluginMetaSet representing the discovered potential-plugins.
+//
+// Currently this supports two different naming schemes. The current
+// standard naming scheme is a subdirectory called $GOOS-$GOARCH containing
+// files named terraform-$KIND-$NAME-V$VERSION. The legacy naming scheme is
+// files directly in the given directory whose names are like
+// terraform-$KIND-$NAME.
```

Comment:
My intent was that this begins a deprecation cycle that leads to versions becoming mandatory in a later major release, since this trick of defaulting unversioned providers to `0.0.0` is done under the assumption that unversioned plugins are always in some sense "older" than versioned ones, because they would've been installed before we had support for versioned providers.